server/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -567,7 +568,11 @@ static void checkBundleJarHell(Set<URL> classpath, Bundle bundle, Map<String, Se
             Set<URL> union = new HashSet<>(classpath);
             union.addAll(bundle.urls);
             JarHell.checkJarHell(union, logger::debug);
-        } catch (Exception e) {
+        } catch (final IOException ioe) {
+            throw new IllegalStateException("failed to load plugin " + bundle.plugin.getName() + " due to io error, caused when checking for Jar Hell", ioe);
+        } catch (final URISyntaxException urie) {
+            throw new IllegalStateException("failed to load plugin " + bundle.plugin.getName() + " due to malformed uri, caused when checking for Jar Hell", urie);
+        } catch (final Exception e) {
```

Comment:
I think the original problem could still occur for another unknown exception. Instead of specializing other exception types, let’s specialize the only one we throw for jar hell, IllegalStateException. Then the general exception can catch all the others with a different message like “failed to load plugin foo while checking for jar hell”.

Reply:
Yes, it makes sense. There may be other exceptions like NPE that might occur. I have made the changes to check only for IllegalStateException that gets specifically thrown, the  general exception catches all the others